pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,569 @@
+// features are backported in various distros and this could get pretty hairy.
+// However iptables-1.4.0 was released 2007-Dec-22 and appears to have every feature we use,
+// so this seems prefectly reasonable for now.
+const (
+	IPTABLES_MIN_VERSION string = "1.4.0"
+)
+
+// the services chain
+var iptablesServicesChain utiliptables.Chain = "KUBE-SERVICES"
+
+// ShouldUseIptablesProxier returns true if we should use the iptables Proxier instead of
+// the userspace Proxier.
+// This is determined by the iptables version. It may return an erorr if it fails to get the
+// itpables version without error, in which case it will also return false.
+func ShouldUseIptablesProxier() (bool, error) {
+	exec := utilexec.New()
+	minVersion, err := semver.NewVersion(IPTABLES_MIN_VERSION)
+	if err != nil {
+		return false, err
+	}
+	versionString, err := utiliptables.GetIptablesVersionString(exec)
+	if err != nil {
+		return false, err
+	}
+	version, err := semver.NewVersion(versionString)
+	if err != nil {
+		return false, err
+	}
+	return !version.LessThan(*minVersion), nil
+}
+
+type portal struct {
+	ip       net.IP
+	port     int
+	protocol api.Protocol
+}
+
+// internal struct for string service information
+type serviceInfo struct {
+	portal              portal
+	loadBalancerStatus  api.LoadBalancerStatus
+	sessionAffinityType api.ServiceAffinity
+	stickyMaxAgeMinutes int
+	endpoints           []string
+	// Deprecated, but required for back-compat (including e2e)
+	deprecatedPublicIPs []string
+}
+
+// returns a new serviceInfo struct
+func newServiceInfo(service proxy.ServicePortName) *serviceInfo {
+	return &serviceInfo{
+		sessionAffinityType: api.ServiceAffinityNone, // default
+		stickyMaxAgeMinutes: 180,                     // TODO: paramaterize this in the API.
+	}
+}
+
+// Proxier is an iptables based proxy for connections between a localhost:lport
+// and services that provide the actual backends.
+type Proxier struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[proxy.ServicePortName]*serviceInfo
+	iptables   utiliptables.Interface
+}
+
+// Proxier implements ProxyProvider
+var _ proxy.ProxyProvider = &Proxier{}
+
+// NewProxier returns a new Proxier given an iptables Interface instance.
+// Because of the iptables logic, it is assumed that there is only a single Proxier active on a machine.
+// An error will be returned if iptables fails to update or acquire the initial lock.
+// Once a proxier is created, it will keep iptables up to date in the background and
+// will not terminate if a particular iptables call fails.
+func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
+	glog.V(2).Info("Tearing down userspace rules.")
+	// remove iptables rules/chains from the userspace Proxier
+	tearDownUserspaceIptables(ipt)
+	return &Proxier{
+		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
+		iptables:   ipt,
+	}, nil
+}
+
+// Chains from the userspace proxy
```

Comment:
will add one next commit. any other todos we should add? EG nodePort, route
to self?

On Tue, Aug 11, 2015 at 3:53 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36790369>
> :
>
> > +// NewProxier returns a new Proxier given an iptables Interface instance.
> > +// Because of the iptables logic, it is assumed that there is only a single Proxier active on a machine.
> > +// An error will be returned if iptables fails to update or acquire the initial lock.
> > +// Once a proxier is created, it will keep iptables up to date in the background and
> > +// will not terminate if a particular iptables call fails.
> > +func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
> > +	glog.V(2).Info("Tearing down userspace rules.")
> > +	// remove iptables rules/chains from the userspace Proxier
> > +	tearDownUserspaceIptables(ipt)
> > +	return &Proxier{
> > +		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
> > +		iptables:   ipt,
> > +	}, nil
> > +}
> > +
> > +// Chains from the userspace proxy
>
> TODO: Remove all of this when the userspace proxier is removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36790369>
> .
>